vital/tests/test_mesh.cxx

```diff
@@ -0,0 +1,236 @@
+// This file is part of KWIVER, and is distributed under the
+// OSI-approved BSD 3-Clause License. See top-level LICENSE file or
+// https://github.com/Kitware/kwiver/blob/master/LICENSE for details.
+
+/// \file
+/// \brief test core mesh functionality
+
+#include <tests/test_gtest.h>
+#include <tests/test_scene.h>
+#include <vital/exceptions.h>
+
+#include <vital/io/mesh_io.h>
+
+#include <gtest/gtest.h>
+
+kwiver::vital::path_t g_data_dir;
+
+// ----------------------------------------------------------------------------
+int main(int argc, char** argv)
+{
+  ::testing::InitGoogleTest( &argc, argv );
+  return RUN_ALL_TESTS();
+}
+
+// ----------------------------------------------------------------------------
+void compare_meshes(const kwiver::vital::mesh_sptr& first,
+                    const kwiver::vital::mesh_sptr& second)
+{
+  kwiver::vital::mesh_vertex_array<3>& first_vertices = dynamic_cast
+    <kwiver::vital::mesh_vertex_array<3>&>( first->vertices() );
+  kwiver::vital::mesh_vertex_array<3>& second_vertices = dynamic_cast
+    <kwiver::vital::mesh_vertex_array<3>&>( second->vertices() );
+
+  EXPECT_EQ( first_vertices.size(), second_vertices.size() );
+  for(unsigned int i=0; i<first_vertices.size(); i++)
+  {
+    EXPECT_EQ( first_vertices[i].size(), second_vertices[i].size() );
+    for(unsigned int j=0; j<first_vertices[i].size(); j++)
+    {
+      EXPECT_EQ( first_vertices[i][j], second_vertices[i][j] );
+    }
+  }
+
+  kwiver::vital::mesh_face_array first_faces = *std::make_shared
+    <kwiver::vital::mesh_face_array>(first->faces());
+  kwiver::vital::mesh_face_array second_faces = *std::make_shared
+    <kwiver::vital::mesh_face_array>(second->faces());
+
+  EXPECT_EQ( first_faces.size(), second_faces.size() );
+  for(unsigned int i=0; i<first_faces.size(); i++)
+  {
+    EXPECT_EQ( first_faces[i].size(), second_faces[i].size() );
+    for(unsigned int j=0; j<first_faces[i].size(); j++)
+    {
+      EXPECT_EQ( first_faces[i][j], second_faces[i][j] );
+    }
+  }
+}
+
+// ----------------------------------------------------------------------------
+TEST(mesh, group_names)
+{
+  kwiver::vital::mesh_sptr cube_mesh = kwiver::testing::cube_mesh( 1.0 );
+  std::shared_ptr<kwiver::vital::mesh_face_array> faces =
+    std::make_shared<kwiver::vital::mesh_face_array>(cube_mesh->faces());
+
+  for(unsigned int i=0; i<faces->size(); i++){
+    EXPECT_EQ( faces->group_name(i), "" );
+  }
+
+  EXPECT_EQ( faces->make_group( "testing name" ), faces->size() );
```

Comment:
This test code works and provides code coverage, but does not fully test that the groups are working.  The idea behind groups is that you first make a group, then add some faces which become part of that group, then make a new group, then add more faces that now belong to the second group, and so on.  Then you should be able to access the indices of the faces that were assigned to each group.  The current API has no easy way to partition an existing set of faces into multiple group without copying those faces into a new face array.

To test you could make a group, add 4 faces, make a second group, add 2 more faces.  Then test that the first group has the 4 expected faces and the second groups and the 2 expected faces.  Or something like that.

Reply:
I expanded the current test to try and capture what you're saying here, by adding faces to a new group and verifying both groups kept the correct sizes. Let me know if it needs any more changes